versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -492,6 +520,156 @@ protected List<RefLog> doFetchPageFromRefLog(
+    doc.put(VALUE_VERSION_PROPERTY_NAME, value.getVersion());
+
+    Document id = doc.get(ID_PROPERTY_NAME, Document.class);
+
+    if (expectedVersion.isPresent()) {
+      UpdateResult r =
+          client
+              .getAttachments()
+              .updateOne(
+                  Filters.and(
+                      Filters.eq(id),
+                      Filters.eq(VALUE_VERSION_PROPERTY_NAME, expectedVersion.get())),
+                  new Document("$set", doc));
+      if (!verifySuccessfulUpdate(r, client.getAttachments())) {
+        return false;
+      }
+    } else {
+      try {
+        if (!client.getAttachments().insertOne(doc).wasAcknowledged()) {
+          return false;
+        }
+      } catch (MongoWriteException writeException) {
+        ErrorCategory category = writeException.getError().getCategory();
+        if (ErrorCategory.DUPLICATE_KEY.equals(category)) {
+          return false;
+        }
+        throw writeException;
+      }
+    }
+
+    verifyAcknowledged(
+        client
+            .getAttachmentKeys()
+            .updateOne(
+                Filters.eq(ID_PROPERTY_NAME, toId(key.getContentId().getId())),
+                Updates.addToSet(DATA_PROPERTY_NAME, attachmentKeyAsString(key)),
+                new UpdateOptions().upsert(true)),
+        client.getAttachmentKeys());
+
+    return true;
+  }
+
+  @Override
+  protected void purgeAttachments(Stream<AttachmentKey> keys) {
+    Map<String, List<AttachmentKey>> attachmentsKeys = new HashMap<>();
+
+    delete(
+        client.getAttachments(),
+        keys.peek(
+                key ->
+                    attachmentsKeys
+                        .computeIfAbsent(key.getContentId().getId(), c -> new ArrayList<>())
+                        .add(key))
+            .map(ProtoSerialization::attachmentKeyAsString));
+
+    attachmentsKeys.forEach(
+        (cid, attachmentKeys) ->
+            verifyAcknowledged(
+                client
+                    .getAttachmentKeys()
+                    .updateOne(
+                        Filters.eq(ID_PROPERTY_NAME, toId(cid)),
+                        Updates.pullAll(
+                            DATA_PROPERTY_NAME,
+                            attachmentKeys.stream()
+                                .map(ProtoSerialization::attachmentKeyAsString)
+                                .collect(Collectors.toList())),
+                        new UpdateOptions().upsert(true)),
+                client.getAttachmentKeys()));
+  }
+
+  @Override
+  protected Stream<AttachmentKey> fetchAttachmentKeys(String contentId) {
+    Document doc =
+        client.getAttachmentKeys().find(Filters.eq(ID_PROPERTY_NAME, toId(contentId))).first();
+    if (doc == null) {
+      return Stream.empty();
+    }
+    List<String> keys = doc.getList(DATA_PROPERTY_NAME, String.class);
+    return keys.stream().distinct().map(ProtoSerialization::attachmentKeyFromString);
```

Comment:
Do we really need `.distinct()` since we use Mongo set operations for writing this data?